### PR TITLE
Initial working commit

### PR DESCRIPTION
**Pull request by SlothOfAnarchy**
- clones and installs the specified version of zfs-snap-manager
- creates the configuration file and restarts the zfs-snap-manager systemd service

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,39 +1,85 @@
...
 ## Requirements
 
-Any pre-requisites that may not be covered by Ansible itself or the role should be mentioned here. For instance, if the role uses the EC2 module, it may be a good idea to mention in this section that the boto package is required.
-
+Obviously, any ZFS filesystems that should be snapshotted must exist.
 
 ## Role Variables
 
-A description of the settable variables for this role should go here, including any variables that are in defaults/main.yml, vars/main.yml, and any variables that can/should be set via parameters to the role. Any variables that are read from other roles and/or the global scope (ie. hostvars, group vars, etc.) should be mentioned here as well.
+| Name                  | mandatory | default |Description                                                                            |
```
Comment (dasJ) [inline README.md]: Indent this table and capitalise mandatory

File: README.md [hunk 2]
```diff
@@ -1,39 +1,85 @@
...
+|-----------------------|--------|----------|-------------------------------------------------------------------------------|
+| `zfs_snap_manager_version` | no | `0.1.13` | [zfs-snap-manager](https://github.com/khenderick/zfs-snap-manager) version to use    |
+| `zfs_snap_manager_clone_dir`  | no | $ansible_env.HOME/.cache || cache directory to clone git repo to                        |
+| `zfs_parent_fs`          | no | `tank` | Parent pool or filesystem (prepended to all datasets)               |
+| `zfs_snapshot_defaults`      | no | {} | Default snapshot configuration used when there is no explicit configuration for a filesystem, see [snapshot configuration](#snapshot configuration) for content.               |
```
Comment (dasJ) [inline README.md]: Make {} a blockquote

File: tasks/install.yml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  apt:
+    name: git
+  when: ansible_pkg_mgr == "apt"
+
+- name: Install dependencies (pacman)
+  pacman:
+    name: git
+  when: ansible_pkg_mgr == "pacman"
+
+- name: Create clone directory
```
Comment (dasJ) [inline tasks/install.yml]: Clone on the control machine instead of the target

**Review by dasJ — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (SlothOfAnarchy) [reply README.md]: I already did !?

**Comment by SlothOfAnarchy** [comment]
please rereview

**Review by dasJ — CHANGES_REQUESTED** [review]
Stop deleting my comments. Clone on the control machine

File: tasks/install.yml [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+- name: Clone zfs-snap-manager
+  git:
+    repo: "https://github.com/khenderick/zfs-snap-manager"
+    dest: "{{ zfs_snap_manager_clone_dir }}/python/src/github.com/khenderick/zfs-snap-manager"
+    version: "v{{ zfs_snap_manager_version }}"
+    force: true
+  register: git
+
+- name: create library directory
```
Comment (dasJ) [inline tasks/install.yml]: Capitalize

File: tasks/install.yml (hunk 3 above)
Comment (dasJ) [inline tasks/install.yml]: Do you really need to do this?

File: tasks/install.yml [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    path: /usr/lib/zfs-snap-manager
+    state: directory
+    owner: root
+    group: root
+    mode: 0755
+  when: git.changed
+
+- name: Copy systemd service
+  copy:
+    src: "{{ zfs_snap_manager_clone_dir }}/python/src/github.com/khenderick/zfs-snap-manager/system/zfs-snap-manager.service"
```
Comment (dasJ) [inline tasks/install.yml]: Use a variable for this long path, it looks awfully copied down

**Review by dasJ — CHANGES_REQUESTED** [review]

File: tasks/install.yml (hunk 3 above)
Comment (SlothOfAnarchy) [reply tasks/install.yml]: no.

**Comment by SlothOfAnarchy** [comment]
@dasJ please rereview and take #2 into consideration.

File: README.md [hunk 6]
```diff
@@ -1,39 +1,85 @@
...
 ## Requirements
 
-Any pre-requisites that may not be covered by Ansible itself or the role should be mentioned here. For instance, if the role uses the EC2 module, it may be a good idea to mention in this section that the boto package is required.
-
+Obviously, any ZFS filesystems that should be snapshotted must exist.
 
 ## Role Variables
 
-A description of the settable variables for this role should go here, including any variables that are in defaults/main.yml, vars/main.yml, and any variables that can/should be set via parameters to the role. Any variables that are read from other roles and/or the global scope (ie. hostvars, group vars, etc.) should be mentioned here as well.
+| Name                         | mandatory | default                  | Description                                                                                                                                                               |
```
Comment (dasJ) [inline README.md]: Fix the indentation and capitalize The headings

File: README.md [hunk 7]
```diff
@@ -1,39 +1,85 @@
...
-A brief description of the role goes here.
+# zfs-snap-manager
 
+This role installs and configures [zfs-snap-manager](https://github.com/khenderick/zfs-snap-manager)
 
 ## Requirements
 
-Any pre-requisites that may not be covered by Ansible itself or the role should be mentioned here. For instance, if the role uses the EC2 module, it may be a good idea to mention in this section that the boto package is required.
-
+Obviously, any ZFS filesystems that should be snapshotted must exist.
```
Comment (dasJ) [inline README.md]: Is a special distribution required?

File: README.md [hunk 8]
```diff
@@ -1,39 +1,85 @@
...
+| `zfs_snapshot_defaults`      | no        | `{}`                     | Default snapshot configuration used when there is no explicit configuration for a filesystem, see [snapshot configuration](#snapshot configuration) for content.          |                                      
+| `zfs_filesystems`            | no       |     `{}`                     | List of filesystems to be snapshotted. Snapshots can be configured using the dict item `snapshots`. For its content see [snapshot configuration](#snapshot configuration) |
+| `zvols`           | no | `{}` | List of zvols to be snapshotted. Snapshots can be configured using the dict item `snapshots`. For its content see [snapshot configuration](#snapshot configuration)                |
+
+### Snapshot configuration
+
+| Name                   | Default     | Description                                                                                                                                                                                                                                                                                            |
```
Comment (dasJ) [inline README.md]: This table doesn't look good either

File: README.md [hunk 9]
```diff
@@ -1,39 +1,85 @@
...
+| `preexec`              | (omitted)   | A command that will be executed, before snapshot/replication. Should be omitted if nothing should be executed                                                                                                                                                                                          |  
+| `postexec`             | (omitted)   | A command that will be executed, after snapshot/replication, but before the cleanup. Should be omitted if nothing should be executed                                                                                                                                                                   |  
+
+For more information on configuration and constraints thereof, consider reading the documentation at [zfs-snap-manager](https://github.com/khenderick/zfs-snap-manager).
+
+## Example Playbook
 
 ```yml
+zfs_parent_fs: rpool
```
Comment (dasJ) [inline README.md:42]: Why not namespace the variables?

File: defaults/main.yml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+zfs_snap_manager_version: 0.1.13
+global_cache_dir: "{{ ansible_env.HOME }}/.cache"
```
Comment (dasJ) [inline defaults/main.yml]: Please do not use this default, use the `mandatory` filter instead

File: tasks/install.yml [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+  when: ansible_pkg_mgr == "apt"
+
+- name: Install dependencies (pacman)
+  pacman:
+    name: git
+  when: ansible_pkg_mgr == "pacman"
+
+- name: Set clone directory variable
+  set_fact:
+    zfs_snap_manager_clone_dir: "{{ global_cache_dir }}/python/src/github.com/khenderick/zfs-snap-manager"
```
Comment (dasJ) [inline tasks/install.yml]: Why not use `vars/main.yml` so you don't have to use this task?

File: tasks/install.yml [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+  register: git
+  notify: Enable and restart zfs-snap-manager
+
+- name: Create library directory
+  file:
+    path: /usr/lib/zfs-snap-manager
+    state: directory
+    owner: root
+    group: root
+    mode: 0755
```
Comment (dasJ) [inline tasks/install.yml]: Why not `750`?

**Review by dasJ — CHANGES_REQUESTED** [review]

**Comment by dasJ** [comment]
@SlothOfAnarchy Also delegate the clone task to `localhost`

File: README.md (hunk 9 above)
Comment (SlothOfAnarchy) [reply README.md:42]: They are namespaced.

File: README.md (hunk 6 above)
Comment (SlothOfAnarchy) [reply README.md]: I don't think it's all that important.

**Comment by SlothOfAnarchy** [comment]
please rereview @dasJ

**Comment by SlothOfAnarchy** [comment]
local cloning is now included (#2)

File: README.md [hunk 13]
```diff
@@ -1,39 +1,89 @@
...
-A brief description of the role goes here.
+# zfs-snap-manager
 
+This role installs and configures [zfs-snap-manager](https://github.com/khenderick/zfs-snap-manager), a tool to automatically create, replicate and destroy ZFS snapshots.
 
 ## Requirements
 
-Any pre-requisites that may not be covered by Ansible itself or the role should be mentioned here. For instance, if the role uses the EC2 module, it may be a good idea to mention in this section that the boto package is required.
-
+`python2` has to be installed on the target machine. Obviously, any ZFS filesystems that should be snapshotted must exist.
```
Comment (dasJ) [inline README.md]: Why not install it with this role?

File: README.md [hunk 14]
```diff
@@ -1,39 +1,89 @@
...
+The machine that executes the playbook needs to have internet access as well as `git` installed.
```
Comment (dasJ) [inline README.md]: Why don't you just install git? "The control machine needs to have internet access and needs to have git installed if it's not a Debian-based distribution"

File: README.md [hunk 15]
```diff
@@ -1,39 +1,89 @@
...
 
 ## Role Variables
 
-A description of the settable variables for this role should go here, including any variables that are in defaults/main.yml, vars/main.yml, and any variables that can/should be set via parameters to the role. Any variables that are read from other roles and/or the global scope (ie. hostvars, group vars, etc.) should be mentioned here as well.
+| Name                         | Mandatory / Default      | Description                                                                             |
```
Comment (dasJ) [inline README.md:12]: Another unindented table

File: README.md [hunk 16]
```diff
@@ -1,39 +1,89 @@
...
+| `global_cache_dir`           | $ansible_env.HOME/.cache | cache directory to clone git repo to |
+| `zfs_snap_manager_version`   | `0.1.13`                 | [zfs-snap-manager](https://github.com/khenderick/zfs-snap-manager) version to use                                                                                         |
+| `zfs_parent_fs`              |                          | Parent pool or filesystem (optionally prepended to all datasets)                     |
+| `zfs_snapshot_defaults`      | `{}`                     | Default snapshot configuration used when there is no explicit configuration for a filesystem, see [snapshot configuration](#snapshot configuration) for content.          |
+| `zfs_filesystems`            | `[]`                     | List of filesystems to be snapshotted. Snapshots can be configured using the dict item `snapshots`. For its content see [snapshot configuration](#snapshot configuration) |
+| `zvols`                      | `[]`                     | List of zvols to be snapshotted. Snapshots can be configured using the dict item `snapshots`. For its content see [snapshot configuration](#snapshot configuration)                |
```
Comment (dasJ) [inline README.md:19]: Why is `zvols` not prefixed?

File: tasks/install.yml [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+
+- name: Copy systemd service
+  copy:
+    src: "{{ zfs_snap_manager_clone_dir }}/system/zfs-snap-manager.service"
+    dest: /etc/systemd/system/zfs-snap-manager.service
+    force: true
+    mode: 0644
+    owner: root
+    group: root
+  notify: Enable and restart zfs-snap-manager
```
Comment (dasJ) [inline tasks/install.yml]: Reload systemd if this changed

**Review by dasJ — CHANGES_REQUESTED** [review]

File: README.md (hunk 14 above)
Comment (SlothOfAnarchy) [reply README.md]: I can't possible know which obscure distribution the executing machine is running. I don't want to add tasks for all packet managers, especially as `git` it is a very common tool, which in most cases will be used to clone this repository in the first place.

File: README.md (hunk 16 above)
Comment (SlothOfAnarchy) [reply README.md:19]: Is `zvols` a common abbreviation or parameter outside of the ZFS context? I think it is unique enough.

File: README.md (hunk 15 above)
Comment (SlothOfAnarchy) [reply README.md:12]: I don't care.

**Review by dasJ — APPROVED** [review]
I guess

**Pull request merged**